### PR TITLE
Multi-Stage Docker file for Glific Frontend

### PR DESCRIPTION
**Pull request by thundersparkf**
## Summary

The existing docker setup for the backend and frontend applications is very tightly coupled. Through this PR and the one submitted for the [backend(#3277)](https://github.com/glific/glific/pull/3277), we can utilise the docker images of both these applications independently. The current approach is present in [glific-project](https://github.com/glific/glific-project) uses submodules and mounted volumes which can reduce flexibility. 

## Approach

Since this uses a multi-stage Dockerfile, both development and production builds are handled through the same.
- The "base" stage in Dockerfile is used to install dependencies.

#### Development
- The "development" stage initialises a development setup using the base image as the building block.
- As before, this uses a ```yarn dev``` to start the application inside the docker container. However, changes have been made to the ```vite.config.ts``` to alter the certificate paths ensuring that we don't need to change any paths for local development or docker containers. 

- ```open: true``` has been modified to false, as inside docker containers, it is unable to open browsers.
- ```package.json``` has been modified to remove the --host option hardcoded as glific.test. In the case of docker in the development environment, the host value needs to be set to "0.0.0.0" as "127.0.0.1 glific.test" value in ```/etc/hosts``` won't allow us to access the application from outside the docker container.

#### Production
- In the "production" stage, we create an optimised build and then serve it in the "serve" stage using Nginx and SSL. SSL Certificates can be mounted as volumes to be accessible inside the container.
- ```nginx.conf.example``` can be copied into ```nginx.conf``` to update any required values.

### Development
In the project root:
- Create SSL certs using the instructions given in [glific](https://github.com/glific/glific) repo.
- ```cp nginx.conf.example nginx.conf``` and edit any required values
- ```cp .env.example .env```
- ```docker build --target development -t glific-frontend:latest -f docker/Dockerfile .```
- ```docker run -p3000:3000 -v <PATH_TO_SSL_CERTIFICATES>:/app/glific-frontend/certs  glific-frontend:latest ```

> Note: To run this setup without docker, the certificates HAVE to be placed inside a ```certs``` folder in the project root or the path has to be edited in the ```vite.config.ts``` file.

### Production
In the project root:
- ```docker build --target serve -t glific-frontend:latest-prod -f docker/Dockerfile .```
- Create SSL certs using the instructions given in [glific](https://github.com/glific/glific) repo.
- ```docker run -p443:443 -v <PATH_TO_SSL_CERTIFICATES>:/app/glific-frontend/certs  glific-frontend:latest-prod```

## ToDo
If both the backend and frontend docker setup are okayed, then the next step can be to combine them both in a full-fledged single command docker-compose deployment setup.

### CODE REVIEW

**Review by mdshamoon — CHANGES_REQUESTED** [review]

**Comment by kurund** [comment]
@thundersparkf 

We should assume local dev and deployments should be possible with or without the dockerize setup. We abstract a few settings to environment variables.

File: package.json [hunk 1]
```diff
@@ -68,7 +68,7 @@
   },
   "scripts": {
     "setup": "yarn install --force && yarn run floweditor",
-    "dev": "vite --host=glific.test",
+    "dev": "vite",
```
[the remark this answers is not in the document]
Comment (thundersparkf) [reply package.json:71]: Would this be resolved by running yarn dev --host glific.test? I've tried that and I'm getting valid certificate as well?

Screenshot 2024-03-09 at 12 05 38

**Comment by thundersparkf** [comment]
Would it also be possible to abstract the certificate paths to env variables in the vite config? I think that would also help in further automation as multiple environments and domains could be flexible. I think I tried but I'm not really a frontend person so I didn't make much headway. At the moment, the paths are hardcoded and either the names would have to changed while mounting the docker volumes or we have to manually edit the vite.config.ts file.

**Comment by mdshamoon** [comment]
@thundersparkf abstracting the certificate paths to the env seems like a good idea. 
See if this helps https://stackoverflow.com/questions/66389043/how-can-i-use-vite-env-variables-in-vite-config-js

Also lets create another command for starting dev server in docker. Maybe 
docker-dev : vite
and lets keep the dev command as it
dev: vite --host=glific.test

This way it won't effect the current contributors who are using the dev command. Otherwise they have to append --host every time

**Comment by mdshamoon** [comment]
Closing since the PR has been inactive for a month.

**Comment by thundersparkf** [comment]
Hi @mdshamoon, got caught up with some stuff. I've made these changes, can I submit another PR if that's cool?

**Pull request closed without merging**